id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/utils/IdentifiableWalker.scala

```diff
@@ -29,8 +29,15 @@ import uk.ac.wellcome.utils.JsonUtil
  * on a map/object node, it looks to see if the node is Identifiable.
  * If so, it adds an identifier to the node.  Otherwise the document is
  * passed through unmodified.
+ *
+ * This class takes a single parameter: a `generateCanonicalId` callback
+ * which receives a list of `SourceIdentifier` instances, and the ontology
+ * type of the object to be identifier.  This callback should return the
+ * canonical ID.
  */
-object IdentifiableWalker {
+case class IdentifiableWalker(
+  generateCanonicalId: (List[SourceIdentifier], String) => String
```

Comment:
You could also compose this functionality by requiring this function in a base class and applying it via a trait.